From 94aa53f3b5bc4d76b1243b0b0cefb3b0bc59c947 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 8 Sep 2014 13:07:50 -0400
Subject: [tools] perf/top: Add --fields option to specify output fields

Message-id: <1410181752-25631-226-git-send-email-jolsa@redhat.com>
Patchwork-id: 91819
O-Subject: [PATCH RHEL7.1 BZ1134356 225/307] perf top: Add --fields option to specify output fields
Bugzilla: 1134356
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1134356
https://bugzilla.redhat.com/show_bug.cgi?id=1134356

upstream
========
commit 6fe8c26d7ab9fbd6748fc40ca5fea1e3131e7236
Author: Namhyung Kim <namhyung@kernel.org>
Date: Tue Mar 4 11:01:41 2014 +0900

description
===========
The --fields option is to allow user setup output field in any order.
It can receive any sort keys and following (hpp) fields:

  overhead, overhead_sys, overhead_us, sample and period

If guest profiling is enabled, overhead_guest_{sys,us} will be
available too.

More more information, please see previous patch "perf report:
Add -F option to specify output fields"
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/Documentation/perf-top.txt b/tools/perf/Documentation/perf-top.txt
index df86328..dcfa54c 100644
--- a/tools/perf/Documentation/perf-top.txt
+++ b/tools/perf/Documentation/perf-top.txt
@@ -116,6 +116,15 @@ Default is to monitor all CPUS.
  local_weight, abort, in_tx, transaction, overhead, sample, period.
  Please see description of --sort in the perf-report man page.
 
+--fields=::
+ Specify output field - multiple keys can be specified in CSV format.
+ Following fields are available:
+ overhead, overhead_sys, overhead_us, sample and period.
+ Also it can contain any sort key(s).
+
+ By default, every sort keys not specified in --field will be appended
+ automatically.
+
 -n::
 --show-nr-samples::
  Show a column with the number of samples.
diff --git a/tools/perf/builtin-top.c b/tools/perf/builtin-top.c
index 280945b..5b389ce 100644
--- a/tools/perf/builtin-top.c
+++ b/tools/perf/builtin-top.c
@@ -1085,6 +1085,8 @@ int cmd_top(int argc, const char **argv, const char *prefix __maybe_unused)
  OPT_STRING('s', "sort", &sort_order, "key[,key2...]",
      "sort by key(s): pid, comm, dso, symbol, parent, cpu, srcline, ..."
      " Please refer the man page for the complete list."),
+ OPT_STRING(0, "fields", &field_order, "key[,keys...]",
+     "output field(s): overhead, period, sample plus all of sort keys"),
  OPT_BOOLEAN('n', "show-nr-samples", &symbol_conf.show_nr_samples,
       "Show a column with the number of samples"),
  OPT_CALLBACK_NOOPT('g', NULL, &top.record_opts,
@@ -1138,17 +1140,18 @@ int cmd_top(int argc, const char **argv, const char *prefix __maybe_unused)
   usage_with_options(top_usage, options);
 
  sort__mode = SORT_MODE__TOP;
+ /* display thread wants entries to be collapsed in a different tree */
+ sort__need_collapse = 1;
 
  if (setup_sorting() < 0) {
-  parse_options_usage(top_usage, options, "s", 1);
+  if (sort_order)
+   parse_options_usage(top_usage, options, "s", 1);
+  if (field_order)
+   parse_options_usage(sort_order ? NULL : top_usage,
+         options, "fields", 0);
   goto out_delete_evlist;
  }
 
- /* display thread wants entries to be collapsed in a different tree */
- sort__need_collapse = 1;
-
- perf_hpp__init();
-
  if (top.use_stdio)
   use_browser = 0;
  else if (top.use_tui)
-- 
1.7.1